### PR TITLE
Add SpeculativeSampler for XLA-compatible decoding

### LINKED ISSUE #2513: RFC: XLA-Optimized Speculative Decoding via SpeculativeSampler
### Problem
Inference latency for large autoregressive models (e.g., Llama3-90B, Gemma-27B) is memory-bandwidth bound. While **Speculative Decoding** (Leviathan et al.) offers significant speedups (2x-3x) by leveraging smaller draft models, KerasHub currently lacks a native, optimized implementation.

Existing external implementations (like Hugging Face's `assisted_generation`) often rely on Python-side orchestration loops. This introduces significant CPU-GPU synchronization overhead ("Python Ping-Pong"), preventing the logic from being fully compiled into a single XLA graph, which is critical for TPU performance.

### Proposed Solution
I propose implementing Speculative Decoding natively in KerasHub by introducing a `SpeculativeSampler` class.

Unlike standard approaches that modify the `generate()` loop, this approach:
1.  **Encapsulates Logic:** Keeps `CausalLM.generate()` clean by isolating speculation logic within the `Sampler` abstraction.
2.  **Enables XLA Compilation:** Orchestrates the *Draft -> Verify -> Accept* loop entirely within the graph (`jax.jit`/`tf.function`), maximizing hardware utilization.
3.  **Static Shape Compatibility:** Uses fixed-length padding for draft sequences to ensure compatibility with XLA/TPU static shape requirements.

### Proposed API
The API follows the standard Keras pattern: the user defines the "How" (Strategy) during compilation/setup, keeping the runtime `generate` call clean.

```python
# 1. Load models
target_model = keras_hub.models.Llama3CausalLM.from_preset("llama3_8b_en")
draft_model = keras_hub.models.Llama3CausalLM.from_preset("llama3_1b_en")

# 2. Attach the speculative strategy
# This allows us to pre-compile the draft generation graph
target_model.compile(
    sampler=keras_hub.samplers.SpeculativeSampler(
        draft_model=draft_model,
        draft_steps=K,  # e.g., 5
        temperature=1.0,
    )
)

# 3. Fast Inference (Clean API)
target_model.generate("The future of AI is", max_length=128)

### PR DESCRIPTION
**Pull request by Vivek1106-04**
## Description

This PR implements [SpeculativeSampler], a native Keras implementation of **Speculative Decoding** ([Leviathan et al., 2023](https://arxiv.org/abs/2211.17192)).

Speculative decoding accelerates autoregressive inference by using a smaller "draft" model to generate candidate tokens that are verified by the target model in parallel. This unlocks speculative speedups for **TPU, XLA, and JAX/TensorFlow** workflows where external engines like vLLM are not available.

### Key Features

- **XLA & TPU Compatible:** Built with vectorized ops (`slice_update`, `cumprod`, `where`) and static shapes for full graph compilation.
- **vLLM API Alignment:** Uses `num_speculative_tokens` argument matching vLLM standards.
- **Parallel Verification:** Supports custom `verify_next` callable for batch target verification.
- **Fallback Safety:** Gracefully falls back to serial verification if `verify_next` is not provided.

### Algorithm

Draft K tokens (serial, cheap) → Verify K+1 positions (parallel) → Accept matching prefix + bonus token

### Usage

```python
import keras_hub

sampler = keras_hub.samplers.SpeculativeSampler(num_speculative_tokens=5)
target_lm.compile(sampler=sampler)

# In custom generate_step, pass draft_next callback:
output = sampler(
    next=target_next,
    prompt=tokens,
    draft_next=draft_next,
)
```
Fixes #2513

### CODE REVIEW

**Comment by Vivek1106-04** [comment]
@sachinprasadhs ! gentle reminder can you review the pr . Thank you

**Comment by sachinprasadhs** [comment]
@Vivek1106-04 , Thanks for the PR, I will take a look into this soon.
